File: code/scripts/prepare_counts.R
library(dplyr)
library(tidyverse)
library(data.table)
library(bedr)
library(readr)

get_counts <- function(file_name, chrom, start, end){
    coord_str <- paste0(chrom, ':', as.character(start), '-', as.character(end))
    print(coord_str)
    print(file_name)
    gene_df <- tabix(coord_str, file_name, check.chr=FALSE, verbose=FALSE)
    colnames(gene_df) <- c('chrom', 'start', 'end', 'counts')
    counts <- get_region_counts(gene_df, chrom, start, end)
    return(counts)
}

get_region_counts <- function(gene_df, chrom, start, end) {
  coord_list <- c()
  sample_counts <- c()
  
  for (i in 1:nrow(gene_df)) {
    start_ <- as.integer(gene_df$start[i])
    end_ <- as.integer(gene_df$end[i])
    if (i == 1) {
      start_ <- start
    }
    if (i == nrow(gene_df)) {
      end_ <- end
    }
    
    segment_length <- end_ - start_
    
    segment_counts <- as.integer(gene_df$counts[i])
    
    segment_to_add <- rep(segment_counts, segment_length)
    
    sample_counts <- c(sample_counts, segment_to_add)
    coord_list <- c(coord_list, seq(start_, end_-1, by=1))
  }
    
  if (!startsWith(chrom, "chr")){
      chrom <- paste0('chr', chrom)
  }
  
  coord_list <- paste(chrom, coord_list, sep=":")
  
  return(list(sample_counts = sample_counts, coord_list = coord_list))
}

process_and_write <- function(assay, chrom, start, end, region_name){
    
    assay_files <- list.files(paste0('coverage/bed/', assay, '/'))
    sample_list <- unique(gsub("\\.bed.*", "", assay_files))
    
    #df <- data.frame()
    count_list <- list()
    for (sample in sample_list){
        file_name <- paste0('coverage/bed/', assay, '/', sample, '.bed.bgz')
        sample_counts <- get_counts(file_name, chrom, start, end)
        sample_counts_to_add <- sample_counts$sample_counts %>% as.numeric() #%>% rbind(df, .)
        print(paste('processed', file_name))
        count_list[[sample]] = sample_counts_to_add
        #df %>% dim() %>% print()
    }
    
    df <- data.frame(count_list) %>% t()
    
    df %>% dim() %>% print()

    colnames(df) <- sample_counts$coord_list
    rownames(df) <- paste0(sample_list, '.', assay)
                           
    out_name <- paste0("coverage/counts/", assay, "/", region_name, ".csv.gz")

    df %>% as.data.frame() %>% rownames_to_column(., var='Sample_ID') %>% write_csv(., out_name, col_names = TRUE)
}

args = commandArgs(trailingOnly=TRUE)
chrom = args[1]
start = as.integer(args[2])
end = as.integer(args[3])
region = args[4]

chrom_rnaseq <- sub("^chr", "", chrom)

process_and_write('rnaseq', chrom_rnaseq, start, end, region)
process_and_write('atacseq', chrom, start, end, region)
process_and_write('K4me3', chrom, start, end, region)
process_and_write('K4me1', chrom, start, end, region)
process_and_write('K27ac', chrom, start, end, region)
process_and_write('K27me3', chrom, start, end, region)

File: code/scripts/prepare_counts.py
import pandas as pd
import gzip
import os
import numpy as np
import tabix
import sys

def process_and_write(assay, chrom, start, end, region_name):
    output_name = 'coverage/counts/' + assay + '/' + region_name + '.csv.gz'
    file_list = sorted(['coverage/bed/' + assay + '/' + x for x in os.listdir('coverage/bed/' + assay) if ('.tbi' not in x)])
    with gzip.open(output_name, 'wt', compresslevel=6) as fh:
        file_count = 1
        for file_name in file_list:
            coord_str, counts_str = process_sample(assay, file_name, chrom, start, end)
            if file_count == 1:
                fh.write(coord_str)
            fh.write(counts_str)
            file_count += 1
    

def process_sample(assay, file_name, chrom, start, end):
    sample_name = file_name.split('/')[-1].split('.')[0]
    if assay == 'rnaseq':
        chrom_for_tabix = chrom[3:]
    else:
        chrom_for_tabix = chrom
    gene_df = tabix_dataframe(file_name, chrom_for_tabix, start, end)
    coord_str, counts_str = get_counts_string(gene_df, sample_name, assay, chrom, start, end)
    return coord_str, counts_str


def tabix_dataframe(file_name, chrom, start, end):
    print(file_name)
    print(chrom, start, end)
    tb = tabix.open(file_name)
    print(type(chrom))
    print(type(start))
    print(type(end))
    print('just to make sure')
    records = tb.query(chrom, start, end)
    print('error is here')
    gene_df = pd.DataFrame(records, columns = ['chrom', 'start', 'end', 'counts'])
    return gene_df


def get_counts_string(gene_df, sample_name, assay, chrom, start, end):
    coord_list = ['Sample_ID']
    sample_counts = [sample_name + '.' + assay]
    for i, (index, row) in enumerate(gene_df.iterrows()):
        start_ = int(row.start)
        end_ = int(row.end)
        if i == 0:
            start_ = start
        if i == len(gene_df) - 1:
            end_ = end

        segment_length = end_ - start_
        segment_counts = int(row.counts)
        segment_to_add = [str(segment_counts)]*segment_length
        sample_counts += segment_to_add
        coord_list += list(range(start_, end_))

    coord_list = [chrom + ':' + str(x) if (x != 'Sample_ID') else x for x in coord_list]
    coord_str = ','.join(coord_list)+'\n'
    counts_str = ','.join(sample_counts)+'\n'
    
    return coord_str, counts_str

if __name__ == '__main__':
    _, chrom, start, end, region_name = sys.argv
    chrom = str(chrom)
    start = int(start)
    end = int(end)
    for assay in ['rnaseq', 'K4me3', 'K4me1', 'K27ac', 'K27me3', 'atacseq']:
        print('processing ' + assay)
        process_and_write(assay, chrom, start, end, region_name)
        print('finished!')
